src/data/polynomial/cyclotomic_polynomials.lean

```diff
@@ -0,0 +1,367 @@
+  { use X - 1,
+    simp only [hone, cyclotomic'_one K, set.mem_univ, pnat.one_coe, subsemiring.coe_top,
+    eq_self_iff_true, map_X, map_one, ring_hom.coe_of, and_self, map_sub] },
+  let B : polynomial K := ∏ i in nat.proper_divisors k, cyclotomic' i K,
+  have Bmo : B.monic,
+  { apply monic_prod_of_monic,
+    intros i hi,
+    exact (cyclotomic'.monic i K) },
+  have Bint : B ∈ lifts (int.cast_ring_hom K),
+  { refine subsemiring.prod_mem (lifts (int.cast_ring_hom K)) _,
+    intros x hx,
+    have xsmall := (nat.mem_proper_divisors.1 hx).2,
+    obtain ⟨d, hd⟩ := (nat.mem_proper_divisors.1 hx).1,
+    rw [mul_comm] at hd,
+  refine hk x xsmall (is_primitive_root.pow (nat.pos_of_ne_zero hzero) hzeta hd) },
+  replace Bint := lifts_and_degree_eq_and_monic Bint Bmo,
+  obtain ⟨B₁, hB₁, hB₁deg, hB₁mo⟩ := Bint,
+  let Q₁ : polynomial ℤ := (X ^ k - 1) /ₘ B₁,
+  have huniq : 0 + B * (cyclotomic' k K) = (X ^ k - 1 :
+  polynomial K) ∧ (0 : polynomial K).degree < B.degree,
+  { split,
+    { rw [zero_add, mul_comm, ← C_1, (X_pow_sub_one_eq_prod_cycl' hpos hzeta),
+      nat.divisors_eq_proper_divisors_insert_self_of_pos hpos],
+      simp only [true_and, finset.prod_insert, not_lt, nat.mem_proper_divisors, dvd_refl] },
+    rw [degree_zero, bot_lt_iff_ne_bot],
+    intro habs,
+    exact (monic.ne_zero Bmo) (degree_eq_bot.1 habs) },
+  replace huniq := div_mod_by_monic_unique (cyclotomic' k K) (0 : polynomial K) Bmo huniq,
+  simp only [lifts, ring_hom.coe_of, ring_hom.mem_srange],
+  use Q₁,
+  rw [(map_div_by_monic (int.cast_ring_hom K) hB₁mo), hB₁, ← huniq.1],
+  simp only [map_pow, map_X, map_one, map_sub]
+end
+
+/-- If `K` is of characteristic `0` and there is a primitive `n`-th root of unity in `K`,
+then `cyclotomic n K` comes from a unique polynomial with integer coefficients. -/
+lemma unique_int_coeff_of_cycl [char_zero K] {ζ : K} {n : ℕ+} (h : is_primitive_root ζ n) :
+  (∃! (P : polynomial ℤ), map (int.cast_ring_hom K) P = cyclotomic' n K) :=
+begin
+  obtain ⟨P, hP⟩ := int_coeff_of_cycl h,
+  rw exists_unique,
+  use P,
+  split,
+  exact hP.1,
+  intros Q hQ,
+  have mapinj : function.injective (map (int.cast_ring_hom K)),
+  { apply map_injective,
+    simp only [int.cast_injective, int.coe_cast_ring_hom] },
+  rw [function.injective] at mapinj,
+  apply mapinj,
+  rw [hP.1, hQ]
+end
+
+end field
+
+section integers
+
+/-- The `n`-th cyclotomic polynomial with coefficients in `ℤ`. -/
+def cyclotomic : ℕ → polynomial ℤ
+| 0 := 1
+| n@(k + 1) := classical.some (int_coeff_of_cycl
+  (complex.is_primitive_root_exp n (nat.succ_ne_zero k)))
+
+lemma cyclotomic_spec (n : ℕ) : map (int.cast_ring_hom ℂ) (cyclotomic n) =
+cyclotomic' n ℂ ∧ (cyclotomic n).degree = (cyclotomic' n ℂ).degree ∧ (cyclotomic n).monic  :=
+begin
+  by_cases hzero : n = 0,
+  { simp only [cyclotomic, cyclotomic'_zero, hzero, degree_one, monic_one, eq_self_iff_true,
+  map_one, and_self] },
+  obtain ⟨k, rfl⟩ := nat.exists_eq_succ_of_ne_zero hzero,
+  rw cyclotomic,
```

Comment:
This line can be deleted because `exact` works up to definitional equality, and all `rw cyclotomic` does is replaces `cyclotomic` with its definition.